cpp/include/tensorrt_llm/batch_manager/kvCacheUtils.h

```diff
@@ -84,25 +88,52 @@ class BlockRange
         auto& blockManager = mManager->getBlockManager();
         for (auto id : mBlockIds)
         {
-            blockHashes.emplace_back(blockManager.getBlockById(id)->getHash());
+            blockHashes.emplace_back(blockManager.getBlockById(id, mWindowSize)->getHash());
         }
         return blockHashes;
     }
 
     void updatePoolIdx(SizeType32 poolIdx)
     {
-        if (mManager)
+        TLLM_CHECK(mManager);
+        TLLM_CHECK(poolIdx > FIRST_POOL_INDEX);
+        mPool = mManager->getBlockManager().getPrimaryPool(poolIdx);
+        auto const newWindowSize = mManager->getBlockManager().getPoolWindowSize(poolIdx);
+        if (newWindowSize != mWindowSize)
         {
-            mPool = mManager->getBlockManager().getPrimaryPool(poolIdx);
+            mWindowSize = newWindowSize;
+            mBlockIds = mManager->getSequence(mRequestId).getCacheBlockIds(mWindowSize).at(SINGLE_BEAM);
         }
     }
 
     friend class BlockIterator;
 
+private:
+    BlockRange(
+        BaseKVCacheManager const& cacheManager, std::vector<SizeType32> blockIds, LlmRequest::RequestIdType requestId)
+        : mManager(&cacheManager)
+        , mPool(cacheManager.getBlockManager().getPrimaryPool(FIRST_POOL_INDEX))
+        , mWindowSize(firstWindowSize(cacheManager))
+        , mRequestId(requestId)
+        , mBlockIds(std::move(blockIds))
+    {
+    }
+
+    static SizeType32 firstWindowSize(BaseKVCacheManager const& cacheManager)
+    {
+        constexpr SizeType32 FIRST_POOL_IDX = 0;
+        return cacheManager.getBlockManager().getPoolWindowSize(FIRST_POOL_IDX);
+    }
+
 private:
     BaseKVCacheManager const* mManager;
     runtime::ITensor::SharedPtr mPool;
+    SizeType32 mWindowSize;
+    const LlmRequest::RequestIdType mRequestId;
     std::vector<SizeType32> mBlockIds;
+
+    static constexpr SizeType32 SINGLE_BEAM = 0;
+    static constexpr SizeType32 FIRST_POOL_INDEX = 0;
```

Comment:
Constants should have prefix `k` and `kSINGLE_BEAM` should probably be called `kFIRST_BEAM`?
```suggestion
    static constexpr SizeType32 kSINGLE_BEAM = 0;
    static constexpr SizeType32 kFIRST_POOL_INDEX = 0;
```